### DISCUSSION TITLE
Support for newer CUDA capability (e.g. sm_86)

### DISCUSSION THREAD

**Discussion by Leengit** [Project-MONAI/monai-deploy-app-sdk#358] [Q&A] [answered]
I successfully build a docker image with `monai-deploy package` that runs on the computer on which I built it.  However when I try to run the same docker image on a computer with a significantly newer / more powerful GPU, it fails.  It appears that the underlying docker image `nvcr.io/nvidia/pytorch:21.07-py3` uses a version of `CUDA` 11.3 and `torch` that do not support `sm_86`.  Upgrading to `torch==1.12.1` within the docker image that I create (and committing the change to a new image that I then use) does not help.  Despite my attempts with `apt-get`, I have been unable to install a newer version of `CUDA` within the created docker image.

Your help with getting support for an `NVIDIA RTX A5000` would be much appreciated!  The error from running the docker image that I created with monai deploy includes
```text
NVIDIA RTX A5000 with CUDA capability sm_86 is not compatible with the current PyTorch installation.
The current PyTorch install supports CUDA capabilities sm_37 sm_50 sm_60 sm_70.
If you want to use the NVIDIA RTX A5000 GPU with PyTorch, please check the instructions at https://pytorch.org/get-started/locally/
...
  File "~/venv/lung/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 453, in _conv_forward
                            weight, bias, self.stride,
                            _pair(0), self.dilation, self.groups)
        return F.conv2d(input, weight, bias, self.stride,
               ~~~~~~~~ <--- HERE   
                        self.padding, self.dilation, self.groups)
RuntimeError: CUDA error: no kernel image is available for execution on the device
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

**Comment by MMelQin** [comment, marked as answer]
Thanks @Leengit for the issue. The default base image for build MONAI App Package is of 2021 vintage, and as you pointed out, out of date. I previously tried out a base container of 22.xx vintage (with newer version of PyTorch), and found some incompatibility issues with the Python packages.

It is important for the App SDK to keep up with the new releases of the dependencies, so I'll convert this to an issue so that we can schedule and track it.

**Reply by Leengit** [reply]
Thank you for your response and for creating Issue https://github.com/Project-MONAI/monai-deploy-app-sdk/issues/359.

**Reply by vikashg** [reply]
@Leengit I am testing your issue right now. Will post an update soon. Did you try with the docker `nvcr.io/nvidia/pytorch:21.12-py3` That image has worked in the past for me.

**Reply by Leengit** [reply]
I used the `monai-deploy package` command, which chose by itself the base docker image to extend.  I don't know how to override that to use `nvcr.io/nvidia/pytorch:21.12-py3`.

**Reply by vikashg** [reply]
ah I see now the problem. I thought you were running monai deploy inside a container and you were having issues. But I get the problem now. You can pass the base docker image using the following command 
```
monai-deploy package -b nvcr.io/nvidia/pytorch:21.12-py3 my_app --tag my_app:latest -m model.ts
```
Please have a look [here](https://github.com/Project-MONAI/monai-deploy-app-sdk/blob/main/notebooks/tutorials/03_segmentation_app.ipynb)

**Reply by vikashg** [reply]
I tested this just now and the packaging works with both 21.11 and 21.12. I have a A6000 gpu with driver version 515.65.01 and CUDA 11.7

**Reply by vikashg** [reply]
@Leengit Just checking if it is resolved

**Reply by Leengit** [reply]
@vikashg Thank you for following up.  Unfortunately it is not resolved.  I built the package with `monai-deploy package -b nvcr.io/nvidia/pytorch:21.12-py3`.  That didn't work directly when I tried to run it because it was assuming a version of `torch` that was incompatible with my `model.zip` file.  So, I ran the docker image, executed
```shell
pip install --update 'colorama>=0.4.1' 'itk' 'matplotlib' \
    'monai-deploy-app-sdk' 'monai[skimage,tqdm,pillow,transformers]' \
    'networkx>=2.4' 'numpy>=1.21.2,<1.23.0' 'pandas' 'pip' 'setuptools' \
    'torch==1.12.1' 'typeguard>=2.12.1' 'wheel'
```
at a `bash` prompt within the running docker container, exited the bash shell, saved the still running container with `docker commit`, pushed the saved image to `ghcr.io`, pulled it on another computer with a `sm_86` GPU, and tried to use it.  I got the same error message:
```text
NVIDIA RTX A5000 with CUDA capability sm_86 is not compatible with the current PyTorch installation.
The current PyTorch install supports CUDA capabilities sm_37 sm_50 sm_60 sm_70.
If you want to use the NVIDIA RTX A5000 GPU with PyTorch, please check the instructions at https://pytorch.org/get-started/locally/
```

There are a bunch of steps there and maybe I messed up on one of them.  I will try redoing this, but if you see anything obviously wrong with this procedure, please let me know.

**Reply by MMelQin** [reply]
Thanks for looking into the specifics and testing it out.

Since we base the MONAI App Package on NVIDAI PyTorch container image, this published [framework compatibility matrix](https://docs.nvidia.com/deeplearning/frameworks/support-matrix/index.html) is often time useful on finding out the details.

**Reply by Leengit** [reply]
Even with 
```shell
monai-deploy package \
    deploy_app.py \
    -b nvcr.io/nvidia/pytorch:22.09-py3 \
    --tag ghcr.io/my_organization/my_repository/my_image_name:22.09-py3 \
    --model model.zip
```
... and then updating Python packages within the produced docker image as described above, I am getting
```text
NVIDIA RTX A5000 with CUDA capability sm_86 is not compatible with the current PyTorch installation.
The current PyTorch install supports CUDA capabilities sm_37 sm_50 sm_60 sm_70.
If you want to use the NVIDIA RTX A5000 GPU with PyTorch, please check the instructions at https://pytorch.org/get-started/locally/
```
That base docker image should be good enough, so I don't understand what is going wrong here.

Although I am running on a computer with the RTX A5000, I am displaying on a weaker machine via SSH X11 forwarding; could that have something to do with it?

**Reply by vikashg** [reply]
@MMelQin Correct me if I  am wrong. But, i would suggest this. 
1. Check your cuda version by running `nvidia-smi` 
2. Create a clean conda environment and install the correct pytorch version based on the cuda version. The correct version can be found on pytorch.org. 
3. Run the map in the environment following one of the tutorials. 
Only after step 3, try packaging it. This will help a lot.

**Reply by Leengit** [reply]
It does work if I skip my customization step of `pip install` within the docker container!

**Reply by MMelQin** [reply]
The PR #381 has been created to fully address the issue here as well as other few minor issues in the Packager.
